handler/OpenIDHandler.php

```diff
@@ -0,0 +1,464 @@
+    {
+        $token = null;
+        if (isset($providerList) && key_exists($selectedProvider, $providerList)) {
+            $settings = $providerList[$selectedProvider];
+            $httpClient = Application::get()->getHttpClient();
+            $response = null;
+            $params = [
+                'code' => $authorizationCode,
+                'grant_type' => 'authorization_code',
+                'client_id' => $settings['clientId'],
+                'client_secret' => $settings['clientSecret'],
+            ];
+            if ($selectedProvider != 'microsoft') {
+                $params['redirect_uri'] = Application::get()->getRequest()->url(
+                    null,
+                    'openid',
+                    'doAuthentication',
+                    null,
+                    array('provider' => $selectedProvider)
+                );
+            }
+            try {
+                $response = $httpClient->request(
+                    'POST',
+                    $settings['tokenUrl'],
+                    [
+                        'headers' => [
+                            'Accept' => 'application/json',
+                        ],
+                        'form_params' => $params,
+                    ]
+                );
+                if ($response->getStatusCode() != 200) {
+                    error_log('Guzzle Response != 200: '.$response->getStatusCode());
+                } else {
+                    $result = $response->getBody()->getContents();
+                    if (isset($result) && !empty($result)) {
+                        $result = json_decode($result, true);
+                        if (is_array($result) && !empty($result) && key_exists('access_token', $result)) {
+                            $token = [
+                                'access_token' => $result['access_token'],
+                                'id_token' => key_exists('id_token', $result) ? $result['id_token'] : null,
+                                'refresh_token' => key_exists('refresh_token', $result) ? $result['refresh_token'] : null,
+                            ];
+                        }
+                    }
+                }
+            } catch (GuzzleException $e) {
+                error_log('Guzzle Exception thrown: '.$e->getMessage());
+            }
+        }
+
+        return $token;
+    }
+
+    /**
+     * This function uses the certs endpoint of the openid provider to get the server certificate.
+     * There are provider-specific differences in case of the certificate.
+     *
+     * E.g.
+     * - Keycloak uses x5c as certificate format which included the cert.
+     * - Other vendors provide the cert modulus and exponent and the cert has to be created via phpseclib/RSA
+     *
+     * If no key is found, null is returned
+     *
+     * @param array $providerList
+     * @param string $selectedProvider
+     * @return array
```

Comment:
Specify ` : ?array` in the function declaration, and remove these annotations.

Reply:
Removed as part of previous comment